### PR TITLE
Use OrderedDict for raw per-module options to preserve section order

### LINKED ISSUE #3675: Overlapping entries in mypy.ini are processed in nondeterministic order
Suppose you have these entries in `mypy.ini`:
```
[mypy-metaserver.lib.shared_folder.*]
ignore_errors = True
[mypy-metaserver.lib.shared_folder.perms.*]
ignore_errors = False
```
For a submodule of package `metaserver.lib.shared_folder.perms` it will be nondeterministic whether that file is processed with `ignore_errors = True` or `ignore_errors = False`.

The reason is that `Options.clone_for_module()` in `mypy/options.py` loops over `self.per_module_options` which is a dict, and dict ordering is randomized by default in Python 3 (unless `PYTHONHASHSEED` is set).

### PR DESCRIPTION
**Pull request by gvanrossum**
Fixes #3675.

Note that the problem only occurs on Python 3.5 and earlier, since in Python 3.6 the standard `dict` type preserves order. Also note that this relies on `ConfigParser` preserving order -- which it does (it uses `OrderedDict` for everything by default).

### CODE REVIEW

File: mypy/options.py [hunk 1]
```diff
@@ -119,7 +120,7 @@ def __init__(self) -> None:
         self.plugins = []  # type: List[str]
 
         # Per-module options (raw)
-        self.per_module_options = {}  # type: Dict[Pattern[str], Dict[str, object]]
+        self.per_module_options = collections.OrderedDict()  # type: MutableMapping[Pattern[str], Dict[str, object]]
```
Comment (ilinum) [inline mypy/options.py]: What's the reason to use `MutableMapping`?
`Dict` should still work, right?

**Comment by gvanrossum** [comment]
Yes, Dict still works, but the call sites don't need it -- MutableMapping
is sufficient. Or maybe we should specify OrderedDict in the annotation
too? (Note that for the value type we have to use exactly Dict because of
invariance.)

File: mypy/options.py (hunk 1 above)
Comment (ddfisher) [inline mypy/options.py]: I think we should consider making this of type `OrderedDict`, because that's important for the semantics.

**Comment by gvanrossum** [comment]
Agreed, pushed a fix. Will merge myself once tests pass.

**Pull request merged**